[Enhancement]: Add SVM `storage_limit` parameter

acch opened the pull request:
Adds ability to configure `storage_limit` parameter of SVM.

Closes #302.

Acceptance tests pass:
```shell
$ TF_ACC=1 go test ./internal/provider/svm/svm_resource_test.go -v
=== RUN   TestAccSvmResource
--- PASS: TestAccSvmResource (33.36s)
PASS
ok      command-line-arguments  33.367s
```

Example Terraform Configuration:
```terraform
// sample SVM with storage_limit parameter
resource "netapp-ontap_svm" "example" {
  cx_profile_name = ...
  name            = ...
  ...
  storage_limit   = "4294967296"
}
```

acch committed:
fix: correct syntax of sample `aggregates` parameter

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add `storage_limit` parameter to svm data source

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add `storage_limit` parameter to svm resource

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
chore: sort data models alphabetically

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add `storage_limit` parameter to svm resource example

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: add acceptance tests for svm resource `storage_limit` parameter

Signed-off-by: Achim Christ <achim.christ@sva.de>

carchi8py commented:
@acch  before we can accept this we need you to follow this (https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/CONTRIBUTING.md). I think the only steps missing right now is step 5.

acch commented:
@carchi8py I had signed the NetApp CCLA on Oct 16... please let me know what's missing.

carchi8py commented:
@acch thank i have that now. Look like the email group we were using was not accepting External email so we didn't get a copy. We've fix that and have your CCLA.

carchi8py commented:
We have a major version 2.0 that in the final stages of our legal review. 

We will be adding this in to our 2.1 release which will be the release after that one.

carchi8py commented:
@acch  Happy Holidays, just a head up we will be reviewing this shorting. Sorry for the hold up we were waiting on Security to finish their review of 2.0 before releasing it and that ended up taking significantly longer than we had hope. Now that, that out of the way we are starting to merge thing in again. Right now I'm working on getting all the dependencies updated and then we'll start review these.

Probably will have review the first week of Jan (or end of this week if I have time).

acch commented:
Happy Holidays @carchi8py! Thanks for the update.
I'll be back in the office the first week of Jan., and will check if anything needs rebasing...

Please let me know what I can do to help.

Thanks much for all your efforts — enjoy a few days off! 🎄

chuyich commented:
@acch Since storage_limit is the new field and only supported 9.13 or higher, please update the support platform section on the document of data source and resource under docs.
Also, the version check has to be done. You may check [this](https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/interfaces/snapmirror_policy.go#L171) as a reference.

chuyich reviewed (changes requested):
Please update doc and add version check.

acch committed:
chore: update docs with new supported platform

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
feat: check version prior to querying SVM by name

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
Hello @chuyich, thank you very much for reviewing this PR.

I have updated the docs and added cluster version checks to `GetSvmByName*()`. Acceptance tests (still) pass.
Please let me know if this is acceptable, or what is missing.

Thanks much!

carchi8py commented on docs/data-sources/svm_data_source.md:
```diff
@@ -13,6 +13,7 @@ Retrieves the configuration of SVM
 ## Supported Platforms
 
 * On-prem ONTAP system 9.6 or higher
+  * `storage_limit` attribute supported with ONTAP system 9.13 or higher
```
This needs to be added in the list of attributes on line 37-46

carchi8py commented on docs/data-sources/svms_data_source.md:
```diff
@@ -13,6 +13,7 @@ Retrieves the configuration of SVMs.
 ## Supported Platforms
 
 * On-prem ONTAP system 9.6 or higher
+  * `storage_limit` attribute supported with ONTAP system 9.13 or higher
```
This needs to be added in the list of attributes on line 64-72

carchi8py commented on docs/resources/svm_resource.md:
```diff
@@ -21,11 +21,12 @@ Create/Modify/Delete a SVM
 ## Supported Platforms
 
 * On-prem ONTAP system 9.6 or higher
+  * `storage_limit` attribute supported with ONTAP system 9.13 or higher
```
This needs to be added in the list of attributes on line 47-63

chuyich commented on internal/provider/svm/svm_resource.go:
```diff
@@ -120,6 +122,12 @@ func (r *SvmResource) Schema(ctx context.Context, req resource.SchemaRequest, re
+				MarkdownDescription: "Maximum storage permitted on svm, in bytes",
+				Optional:            true,
+				Computed:            true,
+				Default:             int64default.StaticInt64(0),
```
It is an optional parameter based on the API doc https://docs.netapp.com/us-en/ontap-restapi-9131/ontap/post-svm-svms.html#recommended-optional-properties
The default value is not needed.

acch replied:
Without a default value, the attribute remains "unknown" after the apply - resulting in the following error:

```
╷
│ Error: Provider returned invalid result object after apply
│ 
│ After the apply operation, the provider still indicated an unknown value for netapp-ontap_svm.example.storage_limit. All values must be known after apply, so this is
│ always a bug in the provider and should be reported in the provider's own repository. Terraform will still save the other known object values in the state.
╵
```

I could manually set the value to 0 in `internal/provider/svm/svm_resource.go > Create()`, but isn't a default value in the schema definition easier to read & understand?

chuyich replied:
You may use 				
```
        PlanModifiers: []planmodifier.Int64{
              IntUseStateForUnknown(),
       },
```
ref. https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/storage/storage_aggregate_resource.go#L159 and see how to handle this parameter in creation/read/update. Since it's optional in the API call, the default value does not have to be set.

acch replied:
Understood. I have removed the default value, and `omitempty` removes the parameter from the request body if unset.

I've adapted the same logic as in [StorageVolumeResource.PercentSnapshotSpace](https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/storage/storage_volume_resource.go#L647).
Please let me know if this is acceptable.

chuyich commented on internal/interfaces/svm.go:
```diff
@@ -24,27 +24,29 @@ type SvmDataModelONTAP struct {
-	Aggregates     []map[string]string `mapstructure:"aggregates"`
+	Name           string              `mapstructure:"name,omitempty"`
+	SnapshotPolicy SnapshotPolicy      `mapstructure:"snapshot_policy,omitempty"`
+	Storage        Storage             `mapstructure:"storage"`
```
Since it's optional, it should be set omitempty.

acch replied:
OK, I did mark the optional `storage` and (nested) `limit` params as 'omitempty'. However, there's some code in place to delete empty params from the request body, which I've adapted to... don't think 'omitempty' will make any difference.

https://github.com/acch/terraform-provider-netapp-ontap/blob/ceeb032a1923209353296a28db20468801db729d/internal/interfaces/svm.go#L221-L232

```go
	if setAggrEmpty {
		delete(body, "aggregates")
	}
	if setCommentEmpty {
		delete(body, "comment")
	}
	if setStorageLimitEmpty {
		// delete storage.limit from request body, so that ONTAP uses default value
		if v, ok := body["storage"].(map[string]interface{}); ok {
			delete(v, "limit")
		}
	}
```

chuyich replied:
You may check https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/storage/storage_volume_resource.go#L647

acch replied:
Understood. I have removed the explicit `setStorageLimitEmpty` parameter, and instead adapted the same logic as in [StorageVolumeResource.PercentSnapshotSpace](https://github.com/NetApp/terraform-provider-netapp-ontap/blob/integration/main/internal/provider/storage/storage_volume_resource.go#L647).
Please let me know if this is acceptable.

acch committed:
chore: add `storage_limit` parameter to attribute list

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
fix: mark optional `storage` and `limit` params as 'omitempty'

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
Thanks @carchi8py and @chuyich for your feedback!

I've added the `storage_limit` argument to the docs. I was under the impression that these lists are generated automatically, due to the `schema generated by tfplugindocs` comment... Now understand that they need to be updated manually.

I've also marked the optional params as 'omitempty', and did verify that acceptance tests (still) pass.
Please let me know if this is acceptable, or what is missing.

Thanks much!

carchi8py commented:
@acch sorry for the late response, i was out last week with a small family emergency.

We will need you to rebase all your pull requests. There was a security issue that required us to rewrite the git history that we made last week. Because of this any pull request or Fork that hasn't rebased is going to appear as all files have been modified. Can you please rebase your pull request.

acch commented:
@carchi8py Sure, I just rebased all PR's.

For our internal planning: how much more time do you think will be required to incorporate our changes into a release available from Terraform registry? Is there already a (rough) target for the 2.1 release?

carchi8py commented:
@acch we don't have a rough date for a release. Let me talk to the team.

But we don't have any hard requirement for 2.1 so we should be able to make 2.1 just your 6 pull request once there merge in and anything move everything else to 2.2.

acch commented:
@carchi8py Are we good to merge this PR?

Please let me know what I need to do exactly to have the PRs added to the provider. My understanding is that I've addressed all requested changes... please let me know what's missing.

Many thanks!

acch committed:
fix: remove explicit 'setStorageLimitEmpty' flag to indicate empty storage.limit

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
fix: replace storage.limit default value with workflow from StorageVolumeResource.PercentSnapshotSpace

Signed-off-by: Achim Christ <achim.christ@sva.de>

suhasbshekar reviewed (changes requested):
Can you fix the lint issues in checks below?

suhasbshekar commented:
Can you update CHANGELOG.md file?

Also your branch is behind the integration/main, we added few more checks to the pipeline as it was missing here, since your PR is from a fork of this repo, so could you pull the new changes into your branch and update the PR?

Just for our reference, since we have disabled ACC tests as our testing profiles are down, acc tests are passing for this PR when tested locally.

acch committed:
fix: correct SVM mock tests

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch committed:
chore: update changelog

Signed-off-by: Achim Christ <achim.christ@sva.de>

acch commented:
@suhasbshekar Thanks for your feedback.

- I've added the version param to `internal/interfaces/svm_test.go`, this should silence the linter error
- I've added an entry to `CHANGELOG.md`
- I've rebased my branch to `integration/main` (once again)

Acc tests are passing on my local test system (ONTAP Select running 9.15.1P4):

```shell
$ date && TF_ACC=1 go test ./internal/provider/svm/svm_resource_test.go -v && date
Tue Feb  4 08:24:24 UTC 2025
=== RUN   TestAccSvmResource
--- PASS: TestAccSvmResource (29.70s)
PASS
ok      command-line-arguments  29.714s
Tue Feb  4 08:24:54 UTC 2025
```

Please let me know what else is needed.
Many thanks!

suhasbshekar commented:
Please resolve the conflicts.

acch commented:
@suhasbshekar I've rebased the PR once again.

Can you please let me know when you'll be able to merge these changes?

suhasbshekar commented:
Will start merging this week, probably will have a release by next week
